Added Chorusfruit teleportaion manipulation

mariusdkm opened the pull request:
This adds the ability to manipulate the destination of a chorusfruit teleportation.

(Inspired by [Are Droppers Actually Random?](https://www.youtube.com/watch?v=Vip25Tet4tA))

**Features**:
- New temprule `chorusManipulation`
- New command `cchorus`
- Seed changes from eating can be calculated, even not fully eaten stuff

With the command `cchorus` a teleportation goal can be set.
That goal is either `relative` or `absolute` to the players position.
The goal can either be an `area` or a `block`.
If a `block` as goal is selected there is another option `--perfectly` this ensures,
that the players hitbox doesn't extend over the blocks boundingbox, so that a player can teleport inside a 1x1 area. 

A block goal is very "improbable" so it requiers a lot of items to be thrown(around 1000+) 
but the chance get's increased the higher the player is standing and the more air is above the block.

**Issues**:
- [x] Since any teleportation above the goal area gets mapped down, the upper y bound is currently ignored, to increase the possibility to land on the desired location. One would have to check every block above the location if that is air.
- [x] If a teleportation fails(it isn't possible to teleport the player to the "random" location),
 the games tries to find a new valid location, which changes the random seed. I can't detect if a teleportation fails, so also the seed change.

- [x] All the text is not translatable
- [ ] All items, except chorus fruits candidates to be thrown, maybe it should be possible to add a blacklist or whitelist?

- [ ] **Not going to change**: When selecting a goal and not using a double, the game defaults to the middle of the block( `8 2 3 -> 8.5 2 3.5` ) 

I would like to highlight the goal somehow, but I have no idea how to do that,
maybe it is possible to use the spectral arrow effect, or the structure block area box?
I coundn't find any function that does that easily.

If there are any structural flaws/issues or generally issues with my code, please let me know,
it's my first large pull request...(you can also contact me on Discord, my tag is `NotSomeBot#1648`)

mariusdkm committed:
Added Chorusfruit teleportaion prediction/manipulation

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/TempRules.java:
```diff
@@ -53,6 +54,15 @@ public static void setMaxEnchantItemThrows(int maxEnchantItemThrows) {
+    @Rule(readOnly = true)
+    //If the goal is relative to the player
+    public static boolean chorusRelativeTel;
+    public static Vec3d chorusGoalV1;
+    public static Vec3d chorusGoalV2;
```
Don't put additional variables in `TempRules`. If you can't think of anywhere to put it, you can make a new class inside the `features` package.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java:
```diff
@@ -0,0 +1,120 @@
+        ));
+    }
+
+    public static <T> com.mojang.brigadier.builder.RequiredArgumentBuilder<ServerCommandSource, net.minecraft.command.argument.PosArgument> areaThen(boolean relative) {
```
Use import statements rather than fully qualified class names.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java:
```diff
@@ -0,0 +1,120 @@
+                    .executes(ctx -> setGoal(getVec3(ctx, "pos1"), getVec3(ctx, "pos2"), relative)));
+    }
+
+    public static <T> com.mojang.brigadier.builder.RequiredArgumentBuilder<ServerCommandSource, net.minecraft.command.argument.PosArgument> blockThen(boolean relative) {
```
Import statements

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java:
```diff
@@ -0,0 +1,120 @@
+        final Vec3d GoalV2 = new Vec3d(Math.max(tempGoalV1.getX(), tempGoalV2.getX()), Math.max(tempGoalV1.getY(), tempGoalV2.getY()), Math.max(tempGoalV1.getZ(), tempGoalV2.getZ()));
+
+        return throwItemsUntil(rand -> {
+            if (particleCount != 16)
```
All if statements should have braces after them. There is some old code in clientcommands that doesn't follow this rule, but all new code should adhere to this.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java:
```diff
@@ -0,0 +1,120 @@
+        return throwItemsUntil(rand -> {
+            if (particleCount != 16)
+                //159 - (7-(itemUseTimeLeft/4)) * 18 = 33 + 4.5 * itemUseTimeLeft
+                for (int i = 0; i < 33 + 4.5 * itemUseTimeLeft; i++)
```
Braces

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/command/ChorusCommand.java:
```diff
@@ -0,0 +1,120 @@
+            if (GoalV1.getX() < x && x < GoalV2.getX() &&
+                    GoalV1.getY() < y && //The top coordinate is special... TODO: Check for things above
+                    GoalV1.getZ() < z && z < GoalV2.getZ()) {
+                if (itemUseTimeLeft == 24) // || itemUseTimeLeft == 24
```
Braces

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
+    public static void onEat(ItemStack stack, Vec3d pos, int particleCount, int itemUseTimeLeft) {
+        if(canMaintainPlayerRNG()) {
+            //Every time a person eats, the particles are random, and when finished more particles spawn(16)
+            for(int i = 0; i < particleCount * 3 + 3; i++)
```
Braces

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
+                nextInt();
+            if (TempRules.chorusManipulation && stack.getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
+                ChorusCommand.onEat(pos, particleCount, itemUseTimeLeft);
+                if(particleCount == 16) //Consumption randoms
+                    for (int i = 0; i < 5; i++)
```
Braces

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -110,8 +114,19 @@ public static void onDrink() {
+                    for (int i = 0; i < 5; i++)
+                        nextInt();
+            }
+        } else
```
Braces

Earthcomputer reviewed (changes requested):
Nice feature! Other than the code style:

mariusdkm committed:
Chorus mainpulation, now in extra class.
Teleport locations are now 99% safe.
Removed --perfectly since no longer needed
Thanks alot to @Vonr for helping me.

mariusdkm commented:
The game doesn't check, if a food regenerates the player, this somehow changes the player seed, so here one may expierience inconsistentcies.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -295,7 +316,7 @@ public static boolean throwItem() {
         Map<Item, Integer> itemCounts = new HashMap<>();
         for (Slot slot : slots) {
-            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0) {
+            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0 && !slot.getStack().getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
```
```suggestion
            if (slot.hasStack() && EnchantmentHelper.getLevel(Enchantments.BINDING_CURSE, slot.getStack()) == 0 && slot.getStack().getItem() != Items.CHORUS_FRUIT) {
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -110,8 +113,25 @@ public static void onDrink() {
+                nextInt();
+            }
+
+            if (TempRules.chorusManipulation && stack.getItem().getTranslationKey().equals("item.minecraft.chorus_fruit")) {
```
```suggestion
            if (TempRules.chorusManipulation && stack.getItem() == Items.CHORUS_FRUIT) {
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -267,9 +287,10 @@ public static boolean throwItemsUntil(Predicate<Random> condition, int max) {
         }
-        if (itemsNeeded > max)
+        if (itemsNeeded > max) {
+            sendError(new LiteralText("Condition not possible or would require more than " + max + " item throws"));
```
Don't put an error message in here, it can be called for reasons other than user interaction. If you want an error message, check the return value.

mariusdkm committed:
Changed all text to translatable text.
Formatted code(Added Braces) and fixed some bugs.

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -250,29 +271,40 @@ private static boolean canMaintainPlayerRNG() {
+        if (itemsNeeded > max) {
+            if (sendErrors) {
+                sendError(new TranslatableText("playerManip.throwError", itemsNeeded));
+            }
             return false;
-
+        }
         for (int i = 0; i < itemsNeeded; i++) {
-            if (!throwItem())
+            if (!throwItem()) {
+                if (sendErrors) {
+                    sendError(new TranslatableText("playerManip.notEnoughItems", i, itemsNeeded));
+                }
```
Why do you need these errors now if you are already emitting an error in `ChorusManipulation.onEat`?

mariusdkm replied:
For detailed feedback.
There can be two reasons a teleport fails. And I think for the end user it is helpfull, to know why he isn't able to manipulate his location.

Earthcomputer replied:
Make an enum with an "OK" result and other values for possible types of failures. Something similar is done for enchanting manipulation.

mariusdkm replied:
So, throwItemUntil should return an Enum?

Earthcomputer replied:
Yes

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -22,12 +23,15 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Predicate;
 
+import static net.earthcomputer.clientcommands.command.ClientCommandManager.sendError;
```
Don't use static imports except in commands

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -49,13 +53,13 @@ public static int nextInt() {
         if ((bound & -bound) == bound)
-            return (int) ((bound * (long)next(31)) >> 31);
+            return (int) ((bound * (long) next(31)) >> 31);
 
         int bits, val;
         do {
             bits = next(31);
             val = bits % bound;
-        } while (bits - val + (bound-1) < 0);
+        } while (bits - val + (bound - 1) < 0);
```
Please don't change the formatting in unrelated code

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -198,8 +219,8 @@ public static void onItemDamage(int amount, LivingEntity holder, ItemStack stack
 
-                        if(stack.getDamage() + amount >= stack.getMaxDamage() - 15){
-                            MinecraftClient.getInstance().player.playSound(SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, 10,0.1f);
+                        if (stack.getDamage() + amount >= stack.getMaxDamage() - 15) {
+                            MinecraftClient.getInstance().player.playSound(SoundEvents.ENTITY_EXPERIENCE_ORB_PICKUP, 10, 0.1f);
```
Same here

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -250,29 +271,40 @@ private static boolean canMaintainPlayerRNG() {
         }
     }
 
-
     // ===== UTILITIES ===== //
-
```
Same here

Earthcomputer reviewed (changes requested):
Also, just to confirm, are you going to address your final two checkboxes?

mariusdkm commented:
I won't adress the issue with the location input.
But if you think the other would be a nice thing to have, I could implement that.
So my question to you is, do you think a black or whitelist would be good, and should it be inside tempRule or inside the chorusManipulation?

mariusdkm committed:
Changed return type of throwItemsUntil
Undid formatting in unrelated code

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -337,6 +357,39 @@ public static long getSeed(Random rand) {
+            return this.message;
+        }
+
+        public ThrowItemsState setMessageArgs(Object... args) {
```
Enums shouldn't be mutable. If you need to store any state, use a non-enum class. Sorry, I did not realize you wanted to store state. (Use 4 spaces to indent ofc, I just didn't want to break my spacebar key typing code in github)
```java
public static class ThrowItemsResult {
  private final Type type;
  private final TranslatableText message;

  public ThrowItemsResult(Type type, Object... args) {
    this.type = type;
    this.message = new TranslatableText(type.getTranslationKey(), args);
  }

  public Type getType() {
    return type;
  }

  public TranslatableText getMessage() {
    return message;
  }

  public enum Type {
    NOT_ENOUGH_ITEMS(false, "playerManip.notEnoughItems"),
    NOT_POSSIBLE(false, "playerManip.throwError"),
    UNKNOWN_SEED(false, "commands.cenchant.uncracked"),
    SUCCESS(true, null),
    ;

    private final boolean success;
    private final String translationKey;

    Type(boolean success, String translationKey) {
      this.success = success;
      this.translationKey = translationKey;
    }

    public boolean isSuccess() {
      return success;
    }

    public String getTranslationKey() {
      return translationKey;
    }
  }
}
```

Earthcomputer commented on src/main/java/net/earthcomputer/clientcommands/features/PlayerRandCracker.java:
```diff
@@ -250,32 +269,33 @@ private static boolean canMaintainPlayerRNG() {
 
-
     // ===== UTILITIES ===== //
-
```
This formatting change still needs reverting. This comment is a section heading, and the whitespace is needed for clarity, and to show that the comment does not merely apply to the next method.

Earthcomputer reviewed (changes requested):
Very nearly ready to approve

mariusdkm committed:
Changed the return type of throwItemsUntil

mariusdkm committed:
Added a beautiful line

Earthcomputer reviewed (approved):
Thank you!
